File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/picker_dialog_color/ColorPickerDialogBuilder.java
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_color;

import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment.DialogCallbacks;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment;
import com.github.rooneyandshadows.lightbulb.dialogs.dialog_custom.CustomDialogBuilder;

import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.LifecycleOwner;

public class ColorPickerDialogBuilder extends BaseDialogBuilder<ColorPickerDialog> {
    private BasePickerDialogFragment.SelectionChangedListener<int[]> changedCallback;
    private int[] selection;
    private final ColorPickerAdapter adapter;

    public ColorPickerDialogBuilder(FragmentManager manager, String dialogTag, ColorPickerAdapter adapter) {
        super(manager, dialogTag);
        this.adapter = adapter;
    }

    public ColorPickerDialogBuilder(LifecycleOwner lifecycleOwner, FragmentManager manager, String dialogTag, ColorPickerAdapter adapter) {
        super(lifecycleOwner, manager, dialogTag);
        this.adapter = adapter;
    }

    @Override
    public ColorPickerDialogBuilder withTitle(String title) {
        return (ColorPickerDialogBuilder) super.withTitle(title);
    }

    @Override
    public ColorPickerDialogBuilder withMessage(String message) {
        return (ColorPickerDialogBuilder) super.withMessage(message);
    }

    @Override
    public ColorPickerDialogBuilder withPositiveButton(BaseDialogFragment.DialogButtonConfiguration positiveButtonConfiguration, BaseDialogFragment.DialogButtonClickListener onClickListener) {
        return (ColorPickerDialogBuilder) super.withPositiveButton(positiveButtonConfiguration, onClickListener);
    }

    @Override
    public ColorPickerDialogBuilder withNegativeButton(BaseDialogFragment.DialogButtonConfiguration negativeButtonConfiguration, BaseDialogFragment.DialogButtonClickListener onClickListener) {
        return (ColorPickerDialogBuilder) super.withNegativeButton(negativeButtonConfiguration, onClickListener);
    }

    @Override
    public ColorPickerDialogBuilder withOnCancelListener(BaseDialogFragment.DialogCancelListener listener) {
        return (ColorPickerDialogBuilder) super.withOnCancelListener(listener);
    }

    @Override
    public ColorPickerDialogBuilder withOnShowListener(BaseDialogFragment.DialogShowListener listener) {
        return (ColorPickerDialogBuilder) super.withOnShowListener(listener);
    }

    @Override
    public ColorPickerDialogBuilder withOnHideListener(BaseDialogFragment.DialogHideListener listener) {
        return (ColorPickerDialogBuilder) super.withOnHideListener(listener);
    }

    @Override
    public ColorPickerDialogBuilder withCancelOnClickOutsude(boolean closeOnClickOutside) {
        return (ColorPickerDialogBuilder) super.withCancelOnClickOutsude(closeOnClickOutside);
    }

    @Override
    public ColorPickerDialogBuilder withDialogType(BaseDialogFragment.DialogTypes dialogType) {
        return (ColorPickerDialogBuilder) super.withDialogType(dialogType);
    }

    @Override
    public ColorPickerDialogBuilder withAnimations(BaseDialogFragment.DialogAnimationTypes animation) {
        return (ColorPickerDialogBuilder) super.withAnimations(animation);
    }

    @Override
    public ColorPickerDialogBuilder withDialogCallbacks(DialogCallbacks callbacks) {
        return (ColorPickerDialogBuilder) super.withDialogCallbacks(callbacks);
    }

    public ColorPickerDialogBuilder withSelectionCallback(BasePickerDialogFragment.SelectionChangedListener<int[]> listener) {
        changedCallback = listener;
        return this;
    }

    public ColorPickerDialogBuilder withSelection(int[] selection) {
        this.selection = selection;
        return this;
    }

    @Override
    public ColorPickerDialog buildDialog() {
        ColorPickerDialog colorPickerDialog = (ColorPickerDialog) fragmentManager.findFragmentByTag(dialogTag);
        if (colorPickerDialog == null)
            colorPickerDialog = ColorPickerDialog.newInstance(title, message, positiveButtonConfiguration, negativeButtonConfiguration, cancelableOnClickOutside, animation);
        colorPickerDialog.setLifecycleOwner(dialogLifecycleOwner);
        colorPickerDialog.setDialogCallbacks(dialogCallbacks);
        colorPickerDialog.setFragmentManager(fragmentManager);
        colorPickerDialog.setDialogTag(dialogTag);
        colorPickerDialog.addOnShowListener(onShowListener);
        colorPickerDialog.addOnHideListener(onHideListener);
        colorPickerDialog.addOnCancelListener(onCancelListener);
        colorPickerDialog.addOnNegativeClickListeners(onNegativeClickListener);
        colorPickerDialog.addOnPositiveClickListener(onPositiveClickListener);
        colorPickerDialog.setAdapter(adapter);
        colorPickerDialog.setOnSelectionChangedListener(changedCallback);
        colorPickerDialog.setSelection(selection);
        return colorPickerDialog;
    }
}

File: lightbulb-dialogs/src/main/java/com/github/rooneyandshadows/lightbulb/dialogs/picker_dialog_color/ColorPickerDialogBuilder.kt
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_color

import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.LifecycleOwner
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment.SelectionChangedListener
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.*
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.callbacks.*

@Suppress("unused")
class ColorPickerDialogBuilder @JvmOverloads constructor(
    dialogTag: String,
    dialogParentFragmentManager: FragmentManager,
    dialogLifecycleOwner: LifecycleOwner? = null,
    initialDialogState: Bundle? = null
) : BaseDialogBuilder<ColorPickerDialog>(dialogTag, dialogParentFragmentManager, dialogLifecycleOwner, initialDialogState) {
    private var changedCallback: SelectionChangedListener<IntArray>? = null
    private var selection: IntArray? = null

    @JvmOverloads
    constructor(dialogTag: String, fragment: Fragment, initialDialogState: Bundle? = null) : this(
        dialogTag,
        fragment.childFragmentManager,
        fragment,
        initialDialogState
    )

    @JvmOverloads
    constructor(dialogTag: String, activity: FragmentActivity, initialDialogState: Bundle? = null) : this(
        dialogTag,
        activity.supportFragmentManager,
        activity,
        initialDialogState
    )

    @Override
    override fun setupNonRetainableSettings(dialog: ColorPickerDialog) {
        dialog.apply {
            changedCallback?.apply { addOnSelectionChangedListener(this) }
        }
    }

    @Override
    override fun setupRetainableSettings(dialog: ColorPickerDialog) {
        dialog.apply {
            setSelection(selection)
        }
    }

    @Override
    override fun initializeNewDialog(): ColorPickerDialog {
        return ColorPickerDialog.newInstance()
    }

    @Override
    override fun withTitle(title: String): ColorPickerDialogBuilder {
        return super.withTitle(title) as ColorPickerDialogBuilder
    }

    @Override
    override fun withMessage(message: String): ColorPickerDialogBuilder {
        return super.withMessage(message) as ColorPickerDialogBuilder
    }

    @Override
    override fun withButton(configuration: DialogButton): ColorPickerDialogBuilder {
        return super.withButton(configuration) as ColorPickerDialogBuilder
    }

    @Override
    override fun withOnCancelListener(listener: DialogCancelListener): ColorPickerDialogBuilder {
        return super.withOnCancelListener(listener) as ColorPickerDialogBuilder
    }

    @Override
    override fun withOnShowListener(listener: DialogShowListener): ColorPickerDialogBuilder {
        return super.withOnShowListener(listener) as ColorPickerDialogBuilder
    }

    @Override
    override fun withOnHideListener(listener: DialogHideListener): ColorPickerDialogBuilder {
        return super.withOnHideListener(listener) as ColorPickerDialogBuilder
    }

    @Override
    override fun withCancelOnClickOutside(closeOnClickOutside: Boolean): ColorPickerDialogBuilder {
        return super.withCancelOnClickOutside(closeOnClickOutside) as ColorPickerDialogBuilder
    }

    @Override
    override fun withDialogType(dialogType: DialogTypes): ColorPickerDialogBuilder {
        return super.withDialogType(dialogType) as ColorPickerDialogBuilder
    }

    @Override
    override fun withAnimations(animation: DialogAnimationTypes): ColorPickerDialogBuilder {
        return super.withAnimations(animation) as ColorPickerDialogBuilder
    }

    @Override
    override fun withDialogListeners(listeners: DialogListeners): ColorPickerDialogBuilder {
        return super.withDialogListeners(listeners) as ColorPickerDialogBuilder
    }

    fun withSelectionCallback(listener: SelectionChangedListener<IntArray>): ColorPickerDialogBuilder {
        changedCallback = listener
        return this
    }

    fun withSelection(selection: IntArray): ColorPickerDialogBuilder {
        this.selection = selection
        return this
    }
}